`strategies.timezones`: No mention of `tzdata` requirement

I don't have `tzdata` installed, and while running this:
```python
from hypothesis import given
import hypothesis.strategies as st

@given(st.timezones())
def test_tz(dt):
    pass
```

I'm getting this error:
```
============================= test session starts =============================
platform win32 -- Python 3.9.9, pytest-6.2.3, py-1.10.0, pluggy-0.13.1
rootdir: c:\Users\leshc\flipio\bot
plugins: anyio-3.3.0, hypothesis-6.29.0, pylama-7.7.1, asyncio-0.14.0
collected 1 item

test_hypothesis_dt.py F                                                  [100%]

================================== FAILURES ===================================
___________________________________ test_tz ___________________________________

key = 'UTC'

    def load_tzdata(key):
        import importlib.resources
    
        components = key.split("/")
        package_name = ".".join(["tzdata.zoneinfo"] + components[:-1])
        resource_name = components[-1]
    
        try:
>           return importlib.resources.open_binary(package_name, resource_name)

C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2544.0_x64__qbz5n2kfra8p0\lib\zoneinfo\_common.py:12: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

package = 'tzdata.zoneinfo', resource = 'UTC'

    def open_binary(package: Package, resource: Resource) -> BinaryIO:
        """Return a file-like object opened for binary reading of the resource."""
        resource = _normalize_path(resource)
>       package = _get_package(package)

C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2544.0_x64__qbz5n2kfra8p0\lib\importlib\resources.py:88: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

package = 'tzdata.zoneinfo'

    def _get_package(package) -> ModuleType:
        """Take a package name or module object and return the module.
    
        If a name, the module is imported.  If the resolved module
        object is not a package, raise an exception.
        """
>       module = _resolve(package)

C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2544.0_x64__qbz5n2kfra8p0\lib\importlib\resources.py:49: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

name = 'tzdata.zoneinfo'

    def _resolve(name) -> ModuleType:
        """If name is a string, resolve to a module."""
        if hasattr(name, '__spec__'):
            return name
>       return import_module(name)

C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2544.0_x64__qbz5n2kfra8p0\lib\importlib\resources.py:40: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

name = 'tzdata.zoneinfo', package = None

    def import_module(name, package=None):
        """Import a module.
    
        The 'package' argument is required when performing a relative import. It
        specifies the package to use as the anchor point from which to resolve the
        relative import to an absolute import.
    
        """
        level = 0
        if name.startswith('.'):
            if not package:
                msg = ("the 'package' argument is required to perform a relative "
                       "import for {!r}")
                raise TypeError(msg.format(name))
            for character in name:
                if character != '.':
                    break
                level += 1
>       return _bootstrap._gcd_import(name[level:], package, level)

C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2544.0_x64__qbz5n2kfra8p0\lib\importlib\__init__.py:127: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

name = 'tzdata.zoneinfo', package = None, level = 0

>   ???

<frozen importlib._bootstrap>:1030: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

name = 'tzdata.zoneinfo', import_ = <function _gcd_import at 0x0000022DCEC11310>

>   ???

<frozen importlib._bootstrap>:1007: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

name = 'tzdata.zoneinfo', import_ = <function _gcd_import at 0x0000022DCEC11310>

>   ???

<frozen importlib._bootstrap>:972: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

f = <function _gcd_import at 0x0000022DCEC11310>, args = ('tzdata',), kwds = {}

>   ???

<frozen importlib._bootstrap>:228: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

name = 'tzdata', package = None, level = 0

>   ???

<frozen importlib._bootstrap>:1030: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

name = 'tzdata', import_ = <function _gcd_import at 0x0000022DCEC11310>

>   ???

<frozen importlib._bootstrap>:1007: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

name = 'tzdata', import_ = <function _gcd_import at 0x0000022DCEC11310>

>   ???
E   ModuleNotFoundError: No module named 'tzdata'

<frozen importlib._bootstrap>:984: ModuleNotFoundError

During handling of the above exception, another exception occurred:

    @given(st.timezones())
>   def test_tz(dt):

test_hypothesis_dt.py:5: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

key = 'UTC'

    def load_tzdata(key):
        import importlib.resources
    
        components = key.split("/")
        package_name = ".".join(["tzdata.zoneinfo"] + components[:-1])
        resource_name = components[-1]
    
        try:
            return importlib.resources.open_binary(package_name, resource_name)
        except (ImportError, FileNotFoundError, UnicodeEncodeError):
            # There are three types of exception that can be raised that all amount
            # to "we cannot find this key":
            #
            # ImportError: If package_name doesn't exist (e.g. if tzdata is not
            #   installed, or if there's an error in the folder name like
            #   Amrica/New_York)
            # FileNotFoundError: If resource_name doesn't exist in the package
            #   (e.g. Europe/Krasnoy)
            # UnicodeEncodeError: If package_name or resource_name are not UTF-8,
            #   such as keys containing a surrogate character.
>           raise ZoneInfoNotFoundError(f"No time zone found with key {key}")
E           zoneinfo._common.ZoneInfoNotFoundError: 'No time zone found with key UTC'

C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2544.0_x64__qbz5n2kfra8p0\lib\zoneinfo\_common.py:24: ZoneInfoNotFoundError
- generated xml file: C:\Users\leshc\AppData\Local\Temp\tmp-20104YBtHtTOtDasf.xml -
=========================== short test summary info ===========================
FAILED test_hypothesis_dt.py::test_tz - zoneinfo._common.ZoneInfoNotFoundErro...
============================== 1 failed in 0.70s ==============================
```

The problem went away after installing `tzdata`. However, the docs don't mention that `tzdata` should be installed prior to using `strategies.timezones`.

--- a/hypothesis-python/RELEASE.rst
+++ b/hypothesis-python/RELEASE.rst
@@ -1,0 +1,6 @@
+RELEASE_TYPE: patch
+
+This patch documents :func:`~hypothesis.strategies.timezones`
+`Windows-only requirement <https://docs.python.org/3/library/zoneinfo.html#data-sources>`__
+for the :pypi:`tzdata` package, and ensures that
+``pip install hypothesis[zoneinfo]`` will install the latest version.

--- a/hypothesis-python/setup.py
+++ b/hypothesis-python/setup.py
@@ -71,7 +71,7 @@ extras = {
     # zoneinfo is an odd one: every dependency is conditional, because they're
     # only necessary on old versions of Python or Windows systems.
     "zoneinfo": [
-        "tzdata>=2020.4 ; sys_platform == 'win32'",
+        "tzdata>=2021.5 ; sys_platform == 'win32'",
         "backports.zoneinfo>=0.2.1 ; python_version<'3.9'",
         "importlib_resources>=3.3.0 ; python_version<'3.7'",
     ],

--- a/hypothesis-python/src/hypothesis/extra/_array_helpers.py
+++ b/hypothesis-python/src/hypothesis/extra/_array_helpers.py
@@ -369,7 +369,6 @@ def mutually_broadcastable_shapes(
     * ``base_shape`` is the shape against which all generated shapes can broadcast.
       The default shape is empty, which corresponds to a scalar and thus does
       not constrain broadcasting at all.
-    * ``shape`` is a tuple of integers.
     * ``min_dims`` is the smallest length that the generated shape can possess.
     * ``max_dims`` is the largest length that the generated shape can possess,
       defaulting to ``max(len(shape), min_dims) + 2``.

--- a/hypothesis-python/src/hypothesis/strategies/_internal/datetime.py
+++ b/hypothesis-python/src/hypothesis/strategies/_internal/datetime.py
@@ -396,6 +396,9 @@ def timezone_keys(
         to install the :pypi:`backports.zoneinfo` module on earlier versions, and
         the :pypi:`importlib_resources` backport on Python 3.6.
 
+        `On Windows, you will also need to install the tzdata package
+        <https://docs.python.org/3/library/zoneinfo.html#data-sources>`__.
+
         ``pip install hypothesis[zoneinfo]`` will install these conditional
         dependencies if and only if they are needed.
 
@@ -453,6 +456,9 @@ def timezones(*, no_cache: bool = False) -> SearchStrategy["zoneinfo.ZoneInfo"]:
         The :mod:`python:zoneinfo` module is new in Python 3.9, so you will need
         to install the :pypi:`backports.zoneinfo` module on earlier versions, and
         the :pypi:`importlib_resources` backport on Python 3.6.
+
+        `On Windows, you will also need to install the tzdata package
+        <https://docs.python.org/3/library/zoneinfo.html#data-sources>`__.
 
         ``pip install hypothesis[zoneinfo]`` will install these conditional
         dependencies if and only if they are needed.

--- a/hypothesis-python/src/hypothesis/strategies/_internal/datetime.py
+++ b/hypothesis-python/src/hypothesis/strategies/_internal/datetime.py
@@ -98,7 +98,10 @@ def datetime_does_not_exist(value):
         # 9999, so it should be a very small fraction of possible values.
         return True
 
-    if value.tzinfo.utcoffset != roundtrip.tzinfo.utcoffset:
+    if (
+        value.tzinfo is not roundtrip.tzinfo
+        and value.utcoffset() != roundtrip.utcoffset()
+    ):
         # This only ever occurs during imaginary (i.e. nonexistent) datetimes,
         # and only for pytz timezones which do not follow PEP-495 semantics.
         # (may exclude a few other edge cases, but you should use zoneinfo anyway)

--- a/hypothesis-python/tests/numpy/test_argument_validation.py
+++ b/hypothesis-python/tests/numpy/test_argument_validation.py
@@ -20,6 +20,8 @@ from hypothesis import strategies as st
 from hypothesis.errors import InvalidArgument
 from hypothesis.extra import numpy as nps
 
+from tests.common.utils import checks_deprecated_behaviour
+
 
 def e(a, **kwargs):
     kw = ", ".join(f"{k}={v!r}" for k, v in kwargs.items())
@@ -267,8 +269,6 @@ def e(a, **kwargs):
         e(nps.basic_indices, shape=(0, 0), max_dims=-1),
         e(nps.basic_indices, shape=(0, 0), max_dims=1.0),
         e(nps.basic_indices, shape=(0, 0), min_dims=2, max_dims=1),
-        e(nps.basic_indices, shape=(0, 0), min_dims=50),
-        e(nps.basic_indices, shape=(0, 0), max_dims=50),
         e(nps.basic_indices, shape=(3, 3, 3), max_dims="not an int"),
         e(nps.integer_array_indices, shape=()),
         e(nps.integer_array_indices, shape=(2, 0)),
@@ -278,5 +278,18 @@ def e(a, **kwargs):
     ],
 )
 def test_raise_invalid_argument(function, kwargs):
+    with pytest.raises(InvalidArgument):
+        function(**kwargs).example()
+
+
+@pytest.mark.parametrize(
+    ("function", "kwargs"),
+    [
+        e(nps.basic_indices, shape=(0, 0), min_dims=50),
+        e(nps.basic_indices, shape=(0, 0), max_dims=50),
+    ],
+)
+@checks_deprecated_behaviour
+def test_raise_invalid_argument_deprecated(function, kwargs):
     with pytest.raises(InvalidArgument):
         function(**kwargs).example()

--- a/tooling/src/hypothesistooling/__main__.py
+++ b/tooling/src/hypothesistooling/__main__.py
@@ -318,6 +318,13 @@ def update_vendored_files():
     if fname.read_bytes().splitlines()[1:] != new.splitlines()[1:]:
         fname.write_bytes(new)
 
+    # Always require the latest version of the tzdata package
+    tz_url = "https://pypi.org/pypi/tzdata/json"
+    tzdata_version = requests.get(tz_url).json()["info"]["version"]
+    setup = pathlib.Path(hp.BASE_DIR, "setup.py")
+    new = re.sub(r"tzdata>=(.+?) ", f"tzdata>={tzdata_version} ", setup.read_text())
+    setup.write_text(new)
+
 
 def has_diff(file_or_directory):
     diff = ["git", "diff", "--no-patch", "--exit-code", "--", file_or_directory]

